ambry-replication/src/main/java/com/github/ambry/replication/BackupCheckerThread.java

```diff
@@ -175,6 +205,101 @@ protected void advanceToken(RemoteReplicaInfo remoteReplicaInfo, ExchangeMetadat
     remoteReplicaInfo.setLocalLagFromRemoteInBytes(exchangeMetadataResponse.localLagFromRemoteInBytes);
     // reset stored metadata response for this replica so that we send next request for metadata
     remoteReplicaInfo.setExchangeMetadataResponse(new ExchangeMetadataResponse(ServerErrorCode.No_Error));
+    logReplicationStatus(remoteReplicaInfo, exchangeMetadataResponse);
+  }
+
+  /**
+   * Write to a given file at a given offset
+   * @param seekableByteChannel File descriptor the file to append to
+   * @param text Text to append
+   * @param offset Offset to write to in the file
+   * @return True if append was successful, false otherwise
+   */
+  protected boolean writeToFile(SeekableByteChannel seekableByteChannel, String text, int offset) {
+    try {
+      if (offset >= 0) {
+        seekableByteChannel.position(offset);
+      }
+      seekableByteChannel.write(ByteBuffer.wrap(text.getBytes(StandardCharsets.UTF_8)));
+      return true;
+    } catch (IOException e) {
+      logger.error(e.toString());
+      return false;
+    }
+  }
+
+  /**
+   * Appends to a given file
+   * @param seekableByteChannel File descriptor the file to append to
+   * @param text Text to append
+   * @return True if append was successful, false otherwise
+   */
+  protected boolean appendToFile(SeekableByteChannel seekableByteChannel, String text) {
+    return writeToFile(seekableByteChannel, text, -1);
+  }
+
+  /**
+   * Creates or returns an existing file descriptor for file replicaCheckStatus
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   * @return Return a file descriptor of type SeekableByteChannel
+   */
+  protected SeekableByteChannel getOrCreateStatusFileFd(RemoteReplicaInfo remoteReplicaInfo) {
+    String filePath = basePath
+        + File.separator + remoteReplicaInfo.getReplicaId().getPartitionId().getId()
+        + File.separator + remoteReplicaInfo.getReplicaId().getDataNodeId().getHostname()
+        + File.separator + "replicaCheckStatus";
+    /**
+     * NOTE: Working with these file options has been a pain. They do not work as advertised in documentation.
+     * You need a WRITE or APPEND flag for the file to be created. Just CREATE does not work !
+     * And TRUNCATE_EXIST needs the file to exist !
+     */
+    return getOrCreateReportFd(filePath, EnumSet.of(StandardOpenOption.CREATE, StandardOpenOption.WRITE));
+  }
+
+  /**
+   * Creates or returns an existing file descriptor for file missingKeys
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   * @return Return a file descriptor of type SeekableByteChannel
+   */
+  protected SeekableByteChannel getOrCreateMissingKeysFd(RemoteReplicaInfo remoteReplicaInfo) {
+    String filePath = basePath
```

Comment:
Use `String.join` here.

Reply:
I am not sure if creating so many files on a prod node is a good idea. Why can't we create just one file and make the log string prefixes unique? What will creating multiple new files give us that creating log entries with appropriate prefix not give us?

Reply:
What is the issue with creating files? There is more than enough space on the host. Logging this stuff in the logs means I'll have to later grep and sort it out. Output from multiple threads will be interleaved. Logs will be rotated and garbage collected. And I may have to rerun the whole thing. Or I'll have to transfer logs to my Mac for post processing. I'd like to avoid all of that. I am sure I will be asked to share the results of this effort. And I want to be able to quickly collect it at a moment's notice.

Reply:
Can we somehow move this file creation logic to closed source repo?

Reply:
yes that's better